### PR TITLE
Support for recorded source/mic directivities from DIRPAT database (#259)

### PR DESCRIPTION
**Pull request by fakufaku**
This PR is to integrate the work done by @prerak23 into the master branch:

### Generalize for other SOFA files

The goal is to generalize what has been done for DIRPAT by @prerak23 to all SOFA files.
The target syntax is as follows.
```python
# the file reader object reads the file and optionally performs interpolation
# if the file contains multiple directivities, they are all read
hrtf = MeasuredDirectivityFile(
    path="hrtf_file.sofa",
    fs=fs,
    interp_order=12,
    interp_n_points=1000,
)

# orientations can be provided as rotation matrices
orientation = Rotation3D([colatitude_deg, azimuth_deg], "yz", degrees=True)

# we can then choose which directivities we want from the file
dir_left = hrtf.get_mic_directivity("left", orientation=orientation)
dir_right = hrtf.get_mic_directivity("right", orientation=orientation)
```

### TODO
- [ ] documentation
- [ ] changelog
- [X] Read multiple objects from a single file to be able to simulate mic arrays more smoothly, e.g. eigenmike. This could be implemented by detecting the multiple channels after the call to `get_response` that would return a `(n_image_sources, n_channels, n_taps)` array.
- [X] Implement 3D rotations with Euler angles

Co-authored-by: prerak23 <prerak23@github.com>

Thanks for sending a pull request (PR), we really appreciate that! Before hitting the submit button, we'd be really glad if you could make sure all the following points have been cleared.

Please also refer to the doc on [contributing](https://pyroomacoustics.readthedocs.io/en/pypi-release/contributing.html) for more details. Even if you can't check all the boxes below, do not hesitate to go ahead with the PR and ask for help there.

- [ ] Are there docstrings ? Do they follow the [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard) style ?
- [ ] Have you run the tests by doing `nosetests` or `py.test` at the root of the repo ?
- [ ] Have you checked that the doc builds properly and that any new file has been added to the repo ? How to do that is covered in the [documentation](https://pyroomacoustics.readthedocs.io/en/pypi-release/contributing.html#documentation).
- [ ] Is there a unit test for the proposed code modification ? If the PR addresses an issue, the test should make sure the issue is fixed.
- [ ] Last but not least, did you document the proposed change in the CHANGELOG file ? It should go under "Unreleased".

Happy PR :smiley:

### CODE REVIEW

**Comment by fakufaku** [comment]
@prerak23 
How are you doing?
I have made good progress, and I am feeling good about releasing things soon.
Would you mind taking a look at the current code?
I have also created some documentation [here](https://pyroomacoustics.readthedocs.io/en/dev-dirpat/pyroomacoustics.directivities.html)
(I don't have your new email address, so please send me an email)

File: examples/directivities/plot_directivity_3D.py [hunk 1]
```diff
@@ -1,17 +1,12 @@
...
 
-pattern = DirectivityPattern.HYPERCARDIOID
 orientation = DirectionVector(azimuth=0, colatitude=45, degrees=True)
 
 # create cardioid object
-dir_obj = CardioidFamily(orientation=orientation, pattern_enum=pattern)
+dir_obj = HyperCardioid(orientation=orientation)
 
 # plot
 azimuth = np.linspace(start=0, stop=360, num=361, endpoint=True)
```
Comment (prerak23) [inline examples/directivities/plot_directivity_3D.py:12]: For colatitude, shouldn't the num be equal to 181?

File: examples/directivities/simulate_binaural_recording.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        default=1000,
+        help="Maximum order to use in the spherical harmonics interpolation",
+    )
+    args = parser.parse_args()
+
+    interp_order = 12
+    interp_n_points = 1000
+
+    fs, speech = wavfile.read(args.source)
+    speech = speech * (0.95 / abs(speech).max())
```
Comment (prerak23) [inline examples/directivities/simulate_binaural_recording.py:65]: Are we normalizing speech signals here? Curious to know why?

File: examples/directivities/simulation_with_measured_directivity.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+eigenmike = MeasuredDirectivityFile("EM32_Directivity", fs=16000)
+akg = MeasuredDirectivityFile("AKG_c480_c414_CUBE", fs=16000)
+
+rot_54_73 = Rotation3D([73, 54], "yz", degrees=True)
+
+dir_obj_Dmic = akg.get_mic_directivity("AKG_c414K", orientation=rot_54_73)
+dir_obj_Emic = eigenmike.get_mic_directivity("EM_32_9", orientation=rot_54_73)
+
+dir_obj_Cmic = FigureEight(
```
Comment (prerak23) [inline examples/directivities/simulation_with_measured_directivity.py:89]: Do you think it is good to comment here that the mic response is analytic for this case ?

File: pyroomacoustics/data/sofa/mit_kemar_large_pinna.sofa
Comment (prerak23) [inline pyroomacoustics/data/sofa/mit_kemar_large_pinna.sofa:1]: This seems to be a new addition, with the past code I haven't experimented with HRTF's.

File: pyroomacoustics/acoustics.py [hunk 4]
```diff
@@ -29,9 +29,10 @@
 import math
 
 import numpy as np
+from scipy import signal
 from scipy.fftpack import dct
 from scipy.interpolate import interp1d
-from scipy.signal import butter, fftconvolve, sosfiltfilt
+from scipy.signal import butter, fftconvolve, hilbert, sosfiltfilt
```
Comment (prerak23) [inline pyroomacoustics/acoustics.py:35]: Haven't found the usage of Hilbert transform in this file.

File: examples/directivities/plot_directivity_3D.py (hunk 1 above)
Comment (fakufaku) [reply examples/directivities/plot_directivity_3D.py:12]: Good point, thanks!

File: examples/directivities/simulation_with_measured_directivity.py (hunk 3 above)
Comment (fakufaku) [reply examples/directivities/simulation_with_measured_directivity.py:89]: Good idea, thanks.

File: examples/directivities/simulate_binaural_recording.py (hunk 2 above)
Comment (fakufaku) [reply examples/directivities/simulate_binaural_recording.py:65]: They may have been provided in integer format or have arbitrary scale. It seems sensible to normalize them.

File: pyroomacoustics/acoustics.py (hunk 4 above)
Comment (fakufaku) [reply pyroomacoustics/acoustics.py:35]: Thanks!

File: pyroomacoustics/data/sofa/mit_kemar_large_pinna.sofa
Comment (fakufaku) [reply pyroomacoustics/data/sofa/mit_kemar_large_pinna.sofa:1]: Yes, indeed. I have added HRTF to experiment with different types of SOFA files.

**Pull request merged**